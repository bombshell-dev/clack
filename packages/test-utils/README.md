# `@clack/test-utils`

A bunch of useful test utiltiies for use inside clack.
