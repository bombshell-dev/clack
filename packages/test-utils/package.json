{
  "name": "@clack/test-utils",
  "private": true,
  "version": "1.0.0-alpha.1",
  "type": "module",
  "main": "./dist/index.js",
  "module": "./dist/index.js",
  "exports": {
    ".": {
      "types": "./dist/index.d.ts",
      "default": "./dist/index.js"
    },
    "./package.json": "./package.json"
  },
  "types": "./dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/bombshell-dev/clack.git",
    "directory": "packages/test-utils"
  },
  "bugs": {
    "url": "https://github.com/bombshell-dev/clack/issues"
  },
  "homepage": "https://github.com/bombshell-dev/clack/tree/main/packages/test-utils#readme",
  "files": ["dist", "CHANGELOG.md"],
  "author": {
    "name": "James Garbutt",
    "url": "https://github.com/43081j"
  },
  "license": "MIT",
  "packageManager": "pnpm@9.14.2",
  "scripts": {
    "build": "tsc",
    "prepack": "pnpm build"
  },
  "dependencies": {
  },
  "devDependencies": {
  }
}
