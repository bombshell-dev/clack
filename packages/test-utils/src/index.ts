export { MockReadable } from './mock-readable.js';
export { MockWritable } from './mock-writable.js';

export function nextTick(): Promise<void> {
	return new Promise<void>((resolve) => {
		setTimeout(resolve, 0);
	});
}
