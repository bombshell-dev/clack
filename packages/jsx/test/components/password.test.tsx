import { MockReadable, MockWritable } from '@clack/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';
import { Password } from '../../src/index.js';

describe('Password', () => {
	let input: MockReadable;
	let output: MockWritable;

	beforeEach(() => {
		input = new MockReadable();
		output = new MockWritable();
	});

	test('renders password input', async () => {
		const task = (<Password message="foo" output={output} input={input} />)();

		input.emit('keypress', '', { name: 'return' });

		const result = await task;

		expect(result).to.equal(undefined);
		expect(output.buffer).toMatchSnapshot();
	});

	test('renders user input', async () => {
		const task = (<Password message="foo" output={output} input={input} />)();

		input.emit('keypress', 'a', { name: 'a' });
		input.emit('keypress', 'b', { name: 'b' });
		input.emit('keypress', '', { name: 'return' });

		const result = await task;

		expect(result).to.equal('ab');
		expect(output.buffer).toMatchSnapshot();
	});

	test('can set custom mask', async () => {
		const task = (<Password message="foo" mask="!" output={output} input={input} />)();

		input.emit('keypress', 'a', { name: 'a' });
		input.emit('keypress', 'b', { name: 'b' });
		input.emit('keypress', '', { name: 'return' });

		const result = await task;

		expect(result).to.equal('ab');
		expect(output.buffer).toMatchSnapshot();
	});
});
