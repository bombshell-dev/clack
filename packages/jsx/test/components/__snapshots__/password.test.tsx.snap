// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Password > can set custom mask 1`] = `
[
  "<cursor.hide>",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36mâ”‚[39m  ![7m[8m_[28m[27m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36mâ”‚[39m  !![7m[8m_[28m[27m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m  [2m!![22m",
  "
",
  "<cursor.show>",
]
`;

exports[`Password > renders password input 1`] = `
[
  "<cursor.hide>",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m",
  "
",
  "<cursor.show>",
]
`;

exports[`Password > renders user input 1`] = `
[
  "<cursor.hide>",
  "[90mâ”‚[39m
[36mâ—†[39m  foo
[36mâ”‚[39m  [7m[8m_[28m[27m
[36mâ””[39m
",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36mâ”‚[39m  â–ª[7m[8m_[28m[27m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=2>",
  "<erase.line><cursor.left count=1>",
  "[36mâ”‚[39m  â–ªâ–ª[7m[8m_[28m[27m",
  "<cursor.down count=2>",
  "<cursor.backward count=999><cursor.up count=4>",
  "<cursor.down count=1>",
  "<erase.down>",
  "[32mâ—‡[39m  foo
[90mâ”‚[39m  [2mâ–ªâ–ª[22m",
  "
",
  "<cursor.show>",
]
`;
